compiler/rustc_const_eval/src/const_eval/machine.rs

```diff
@@ -429,14 +430,44 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
     }
 
     fn before_terminator(ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
-        // The step limit has already been hit in a previous call to `before_terminator`.
-        if ecx.machine.steps_remaining == 0 {
-            return Ok(());
-        }
-
-        ecx.machine.steps_remaining -= 1;
-        if ecx.machine.steps_remaining == 0 {
-            throw_exhaust!(StepLimitReached)
+        if let Some(new_steps) = ecx.machine.num_evaluated_terminators.checked_add(1) {
+            ecx.machine.num_evaluated_terminators = new_steps;
+            if new_steps == 1_000_000 {
+                // By default, we stop after a million steps, but the user can disable this lint
+                // to be able to run until the heat death of the universe or power loss, whichever
+                // comes first.
+                let hir_id = ecx.best_lint_scope();
+                let is_error = ecx
+                    .tcx
+                    .lint_level_at_node(
+                        rustc_session::lint::builtin::LONG_RUNNING_CONST_EVAL,
+                        hir_id,
+                    )
+                    .0
+                    .is_error();
+                ecx.tcx.struct_span_lint_hir(
+                    rustc_session::lint::builtin::LONG_RUNNING_CONST_EVAL,
+                    hir_id,
+                    ecx.tcx.span,
+                    "this constant is taking a long time to get evaluated",
+                    |diag| diag,
+                );
+                if is_error {
+                    let guard = ecx
+                        .tcx
+                        .sess
+                        .delay_span_bug(ecx.tcx.span, "The deny lint should have already errored");
+                    throw_inval!(AlreadyReported(guard));
+                }
+            } else if new_steps > 10_000_000 && new_steps.count_ones() == 1 {
```

Comment:
Why not `is_power_of_two`?

Also, another number to please put into a `const`.